### ISSUE TITLE
Lazy Services

### ISSUE THREAD

**Issue by ausi** [contao/contao#4154]
As discussed in Slack, #4151 was a quickfix to make the Contao Manager work with Contao 4.13.0.

We should discuss if we want to keep the `lazy: true` there and therefore use `lazy: true` for other things as well in the future, or if we should go with a different solution instead (and decide to never use `lazy: true`).

What I don’t like about proxy classes is, that they are not visible to the developer who is using them. Writing code like the following will not work as expected if the service is lazy:

```php
function doSomethingRisky(MyService $myService) {
	try {
		$myService->somethingRisky();
	} catch (\Throwable $exception) {
		return false;
	}
	
	return true;
}
```

Such a lazy feature needs to be visible to the developer IMO.  
In an optimal world this would look like this:

```php
function doSomethingRisky(Lazy<MyService> $myServiceProxy) {
	$myService = $myServiceProxy->get();
	
	try {
		$myService->somethingRisky();
	} catch (\Throwable $exception) {
		return false;
	}
	
	return true;
}
```

Currently possible would be something like this:

```php
function doSomethingRisky(MyService&LazyLoadingInterface $myService) {
	$myService->initializeProxy();
	
	try {
		$myService->somethingRisky();
	} catch (\Throwable $exception) {
		return false;
	}
	
	return true;
}
```

**Comment by ausi** [comment]
Searching through Symfony I think Service Closures match best what we want to achieve:

```yaml
contao.listener.command_scheduler:
    //...
    arguments:
        - !service_closure '@contao.cron'
        //...
```

```php
/**
 * @param callable(): Cron $cronLazy
 */
function __construct(\Closure $cronLazy, …) {
	$cron = $cronLazy();
	$cron->run();
	// OR
	$cronLazy()->run();
}
```

**Comment by fritzmg** [comment]
But then you would need to have prior knowledge about the service that you are trying to inject. Whereas with `lazy: true` you don't have to worry about that at all. From a DX point of view I do not want to have to know that some services should be injected lazily and some do not - and I do not want to need to take care of that myself.

**Comment by m-vo** [comment]
> But then you would need to have prior knowledge about the service that you are trying to inject. Whereas with lazy: true you don't have to worry about that at all. 

That's exactly the point. For DI it's crucial to know if you're working with a future of an object or an object. IMHO using transparent ghost objects is not bad in general, but for this scenario it's a very bad design pattern.

**Comment by ausi** [comment]
> But then you would need to have prior knowledge about the service that you are trying to inject.

This can (and IMO should) be solved by making the dependencies of `contao.cron` themself lazy. Should be easily done like so:

```yaml
contao.cron:
    class: Contao\CoreBundle\Cron\Cron
    public: true
    arguments:
        - !service_closure '@contao.repository.cron_job'
        - !service_closure '@doctrine.orm.entity_manager'
        - '@?logger'
```

This has basically the same affect as the current `lazy: true` but without “magic” objects that get handed to the developers.

**Comment by fritzmg** [comment]
I would be more in favour of that then. Btw. wouldn't we have to do this with _all_ services that inject a entity repository or the entity manager?

**Comment by ausi** [comment]
> wouldn't we have to do this with _all_ services that inject a entity repository or the entity manager?

IMO no. Most services that require these need a working database connection anyway. But it might make sense for some services (where the DB access is optional or not used for usual use cases).

**Comment by aschempp** [comment]
I don't understand the problem. The advantage of `lazy: true` is that it is completely transparent to the developer. I expect a service, I get an instance of that service, and the service works as intended. Why would I need to know that the service is lazy or that I need to call a service closure?

**Comment by m-vo** [comment]
For the reason of the initial post? They really are a hacky concept. 🙃 Plus: final classes do not work.

**Comment by ausi** [comment]
> I don't understand the problem.

In my first code example, the try/catch block catches an exception from the service initialization which it isn’t meant to catch. The fact that you didn’t see that problem at first glance is exactly my point. No developer would expect that `$myService->somethingRisky()` actually calls the DI container.

> The advantage of `lazy: true` is that it is completely transparent to the developer.

The fact that the developer using the service doesn’t know about it is a huge disadvantage IMO.

**Comment by aschempp** [comment]
>> For the reason of the initial post? 

> What I don’t like about proxy classes is, that they are not visible to the developer who is using them.

That's exactly the advantage to me. The chance that initialisation fails like in the initial post is probably 1000 times lower than the case we just had, people now knowing/remembering to use the class in a lazy way if they should.

> They really are a hacky concept. 

Really? Everyone uses them in Doctrine entities without issues for years. A proxy (a well-known design pattern) is hacky, but we were [mostly fine with using a lib to bypass `final`](https://github.com/contao/contao/pull/3185)? 🤷 

---

Btw on the `Cron` class problem, we might also just fix that by lazy-fetching the entity repository _inside_ the cron class, couldn't we? 🙃

**Comment by ausi** [comment]
> Btw on the `Cron` class problem, we might also just fix that by lazy-fetching the entity repository _inside_ the cron class, couldn't we? 🙃

You mean what I suggested in https://github.com/contao/contao/issues/4154#issuecomment-1043029668 ?

**Comment by aschempp** [comment]
Ah yes! I misunderstood it to people having to do the `service_closure` if they use the `Cron` service 🎉

**Comment by ausi** [comment]
@contao/developers do you agree to use `!service_closure` for the `contao.cron` service as suggested in https://github.com/contao/contao/issues/4154#issuecomment-1043029668 ? If so, please 👍 and I will create a PR.

**Comment by ausi** [comment]
Closed in favor of #4265
